node/forest_libp2p/src/blocksync/provider.rs

```diff
@@ -0,0 +1,241 @@
+        let mut tipset_bundle: TipsetBundle = Default::default();
+        let tipset = match chain::tipset_from_keys(db.as_ref(), &TipsetKeys::new(curr_tipset_cids))
+        {
+            Ok(tipset) => tipset,
+            Err(err) => {
+                debug!("Cannot get tipset from keys: {}", err);
+
+                return BlockSyncResponse {
+                    chain: vec![],
+                    status: BlockSyncResponseStatus::InternalError,
+                    message: "Tipset was not found in the database".to_owned(),
+                };
+            }
+        };
+
+        if request.include_blocks() {
+            tipset_bundle.blocks = tipset.blocks().to_vec();
+        }
+
+        if request.include_messages() {
+            match compact_messages(db.as_ref(), &tipset) {
+                Ok(compacted_messages) => tipset_bundle.messages = Some(compacted_messages),
+                Err(err) => {
+                    debug!("Cannot compact messages for tipset: {}", err);
+
+                    return BlockSyncResponse {
+                        chain: vec![],
+                        status: BlockSyncResponseStatus::InternalError,
+                        message: "Can not fullfil the request".to_owned(),
+                    };
+                }
+            }
+        }
+
+        response_chain.push(tipset_bundle);
+
+        if response_chain.len() as u64 >= request.request_len || tipset.epoch() == 0 {
+            break;
+        }
+
+        curr_tipset_cids = tipset.parents().cids().to_vec();
+    }
+
+    let result_chain_length = response_chain.len() as u64;
+
+    BlockSyncResponse {
+        chain: response_chain,
+        status: if result_chain_length < request.request_len {
+            BlockSyncResponseStatus::PartialResponse
+        } else {
+            BlockSyncResponseStatus::Success
+        },
+        message: "Success".to_owned(),
+    }
+}
+
+// Builds CompactedMessages for given Tipset.
+fn compact_messages<DB>(db: &DB, tipset: &Tipset) -> Result<CompactedMessages, ChainError>
+where
+    DB: BlockStore,
+{
+    let mut bls_messages_order = HashMap::new();
+    let mut secp_messages_order = HashMap::new();
+    let mut bls_cids_combined: Vec<Cid> = vec![];
+    let mut secp_cids_combined: Vec<Cid> = vec![];
+    let mut bls_msg_includes: Vec<Vec<u64>> = vec![];
+    let mut secp_msg_includes: Vec<Vec<u64>> = vec![];
+
+    for block_header in tipset.blocks().iter() {
+        let (bls_cids, secp_cids) = chain::read_msg_cids(db, block_header.messages())?;
+
+        let mut block_include = vec![];
+        let mut secp_include = vec![];
+
+        for bls_cid in bls_cids.iter() {
+            let order = match bls_messages_order.get(bls_cid) {
+                Some(order) => *order,
+                None => {
+                    let order = block_include.len() as u64;
+                    bls_cids_combined.push(bls_cid.clone());
+                    bls_messages_order.insert(bls_cid.clone(), order);
```

Comment:
Oh and one other thing, this can be `.into_iter()` to move the vector and require one less clone for each. (Nbd though, I can change after also)

Reply:
Good call :+1: Updated. Though I need the first clone, as I'm pushing to `Vec` and inserting to `HashMap` the same value, and hashmap consumes the ownership.